common/OpTestIPMI.py

```diff
@@ -160,64 +161,118 @@ def set_system_to_UNKNOWN_BAD(system):
+        self.sol = None
+        self.delaybeforesend = delaybeforesend
+        self.block_setup_term = block_setup_term # allows caller specific control of when to block setup_term
+        self.setup_term_quiet = 0 # tells setup_term to not throw exceptions, like when system off
+        self.setup_term_disable = 0 # flags the object to abandon setup_term operations, like when system off
+
+        # FUTURE - System Console currently tracked in System Object
+        # state tracking, reset on boot and state changes
+        self.PS1_set = -1
+        self.LOGIN_set = -1
+        self.SUDO_set = -1
 
     def set_system(self, system):
         self.system = system
 
-    def terminate(self):
-        if self.state == IPMIConsoleState.CONNECTED:
-            self.sol.terminate()
-            self.state = IPMIConsoleState.DISCONNECTED
+    def set_system_setup_term(self, flag):
+        self.system.block_setup_term = flag
+
+    def get_system_setup_term(self):
+        return self.system.block_setup_term
+
+    def set_block_setup_term(self, flag):
+        self.block_setup_term = flag
+
+    def get_block_setup_term(self):
+        return self.block_setup_term
+
+    def enable_setup_term_quiet(self):
+        self.setup_term_quiet = 1
+        self.setup_term_disable = 0
+
+    def disable_setup_term_quiet(self):
+        self.setup_term_quiet = 0
+        self.setup_term_disable = 0
 
     def close(self):
+        self.util.clear_state(self)
         if self.state == IPMIConsoleState.DISCONNECTED:
             return
         try:
-            if not self.sol.isalive():
-                print "IPMI SOL Console is already disconnected"
-                pass
             self.sol.send("\r")
             self.sol.send('~.')
-            self.sol.expect('[terminated ipmitool]')
-            self.sol.close()
+            close_rc = self.sol.expect(['[terminated ipmitool]', pexpect.TIMEOUT, pexpect.EOF], timeout=10)
+            rc_child = self.sol.close()
+            self.state = IPMIConsoleState.DISCONNECTED
+            exitCode = signalstatus = None
+            if self.sol.status != -1: # leaving for future debug
+              if os.WIFEXITED(self.sol.status):
+                exitCode = os.WEXITSTATUS(self.sol.status)
+              else:
+                signalstatus = os.WTERMSIG(self.sol.status)
         except pexpect.ExceptionPexpect:
+            self.state = IPMIConsoleState.DISCONNECTED
             raise OpTestError("IPMI: failed to close ipmi console")
-
-        self.sol.terminate()
-        self.state = IPMIConsoleState.DISCONNECTED
+        except Exception as e:
+            self.state = IPMIConsoleState.DISCONNECTED
+            pass
 
     def connect(self):
         if self.state == IPMIConsoleState.CONNECTED:
-            self.sol.terminate()
-            self.state = IPMIConsoleState.DISCONNECTED
+          rc_child = self.close()
+          self.state = IPMIConsoleState.DISCONNECTED
```

Comment:
self.sol.close() already takes care of setting the console state to DISCONNECTED, so in the below three cases you can remove state set to DISCONNECTED line.

Reply:
> remove state set to DISCONNECTED line

Fixed 3 cases.